lib/AST/SubstitutionMap.cpp

```diff
@@ -456,13 +456,26 @@ SubstitutionMap SubstitutionMap::subst(TypeSubstitutionFn subs,
   if (empty()) return SubstitutionMap();
 
   SmallVector<Type, 4> newSubs;
-  for (Type type : getReplacementTypesBuffer()) {
-    if (!type) {
-      // Non-canonical parameter.
-      newSubs.push_back(Type());
-      continue;
+  newSubs.reserve(getReplacementTypesBuffer().size());
+  {
+    bool isIdentitySubstitution = true;
+    for (const auto &oldTy : getReplacementTypesBuffer()) {
+      if (!oldTy) {
+        // Non-canonical parameter.
+        newSubs.push_back(Type());
+        continue;
+      }
+
+      const auto newTy = oldTy.subst(subs, conformances, options);
+      newSubs.push_back(newTy);
+
+      isIdentitySubstitution &= newTy && newTy->isEqual(oldTy);
+    }
+
+    if (isIdentitySubstitution) {
+      // The replacement types have not changed.
```

Comment:
I think it's possible that the conformances change but the replacement types do not, if for example a type parameter is replaced with another type parameter, but a nested type was made concrete under the substitution. See 42a3ad3bf0d95e810c90cbbbe2875f53c502a01f for a related problem.